Fixes to awards-bot

laijasmine opened the pull request:
- Deals with changes in the RT package, resolves #25 
- Fixes issue from: `Error in UseMethod("months") : no applicable method for 'months' applied to an object of class "character"`

Commit:
add to documentation

Commit:
remove rmarkdown experiment

Commit:
fix carriage return to send full email

Commit:
instruction notes and logo

Commit:
remove check_rt_reply

Commit:
make sure the numbers are numeric

Commit:
remove ticket error test for now

Commit:
remove sections where RT package no longer supports

laijasmine commented:
Chris - I think I found the source of the problem that it was reading in the `Sys.getenv` variables as characters instead of numeric. Made some changes.. hopefully we can test this out some time.

csjx commented on R/RT_functions.R:
```diff
@@ -208,20 +203,6 @@ send_correspondence_at_time_x <- function(awards_db,
 }
 
-  
-## helper function to check RT replies
```
@laijasmine - What was the reason for removing `check_rt_reply()`? It seems useful to report to the Slack channel when things aren't working as expected.  Is there a problem on the RT side?  When you can, add a detail paragraph to your commits after the short one-line description so we know what the rationale is for each commit (other than minor formatting or typo fix commits).

laijasmine replied:
Thanks Chris! Will definitely do that going forward. My apologies for having you review this without much context.

This is an issue with the `RT` package. I checked the most recent version here for the [`rt_ticket_history_reply` ](https://github.com/NCEAS/rt/blob/master/R/rt_ticket_history_reply.R) function and it now only returns the ticket id and not all the information as before (ie. `reply$status_code`). Therefore, the `check_rt_reply` no longer works as intended because it is dependent on the other information that is returned in older versions of this function.

laijasmine replied:
tried to fixed in subsequent updates

csjx commented on R/main.R:
```diff
@@ -5,6 +5,12 @@ main <- function(database_path = Sys.getenv('DATABASE_PATH'),
                  annual_report_time = Sys.getenv('ANNUAL_REPORT_TIME'),
                  initial_aon_offset = Sys.getenv('INITIAL_AON_OFFSET'),
                  aon_recurring_interval = Sys.getenv('AON_RECURRING_INTERVAL')) {
+  
```
Looks good 👍 - thanks for fixing the return type.

csjx commented on R/main.R:
```diff
@@ -31,11 +37,17 @@ test_main <- function(database_path = Sys.getenv('DATABASE_PATH'),
   db$pi_email <- email
   utils::write.csv(db, database_path, row.names = FALSE)
   
+  #runs fine individually but has issues when run from the top
```
Is this a bug comment?  If so, let's make a Github issue.

laijasmine replied:
I'll remove it!  This was an issue I had running it locally on my Mac that I made a note to myself when initially was getting to know the bot. I must have committed it by accident.

csjx commented on tests/testthat/test_RT_functions.R:
```diff
@@ -18,20 +18,20 @@ test_that('we can create an RT ticket', {
-  # create_ticket ouputs the character 'error' when it fails
-  expect_equal(ticket, 'rt_ticket_create_error')
-})
+# test_that('create_ticket errors gracefully', {
```
Again, please explain why you're removing the unit test in your commit detail paragraph - it's hard to know what's failing here.

csjx commented on R/RT_functions.R:
```diff
@@ -37,16 +37,6 @@ create_ticket <- function(award, requestor) {
                                  subject = subject,
                                  rt_base = 'https://support.nceas.ucsb.edu/rt')
 
-   # if (!grepl('created', httr::content(ticket))) {
```
Can you explain in your commit what the RT package isn't supporting?

laijasmine replied:
fixed to support the new changes in RT

csjx commented on tests/testthat/test_RT_functions.R:
```diff
@@ -18,21 +18,6 @@ test_that('we can create an RT ticket', {
   expect_type(ticket, 'character')
 })
 
-# test_that('create_ticket errors gracefully', {
```
Same here - if RT is failing these calls, let's add Github tickets rather than removing the work, and we can discuss what to do.

laijasmine replied:
#25

csjx commented on tests/testthat/test_RT_functions.R:
```diff
@@ -10,24 +11,11 @@ test_that('we can create an RT ticket', {
-  }
-  
-  db <- create_dummy_database()
+  #does not return ticket but creates ticket
```
If `create_ticket()` doesn't return the ticket, what does it return? Instead of removing the functionality, can we get the ticket based on what is returned (like the `id`)?

laijasmine replied:
I removed the part of this test that involves the same rt problem with `rt_ticket_history_reply` in the `create_ticket` code where it no longer gives the error code. Without the error code, I can't check if it creates the `rt_ticket_create_error` object. So now I have changed the test to see if a ticket to see if `create_ticket` indeed returns some character object (the ticket `id`)

amoeba replied:
(Jump jumping in here with a note on the change in the `rt` package behavior we introduced for the first release: If `rt_ticket_history_reply` fails, it _should_ throw a `stop` which can be `tryCatch`'d. The error message, if one was sent back by RT, will be available in the `tryCatch`. If it's not, let us know over on https://github.com/NCEAS/rt and we can put in a workaround).

laijasmine replied:
Oh nice I will look into that! I should have asked earlier! Thanks Bryce

csjx replied:
Thanks @amoeba !

csjx reviewed (changes requested):
Heya @laijasmine - Let's chat about these changes, particularly the RT calls not working.  See my comments in the code.  Also, try to be more verbose in your commits (one-line title, followed by a second line or paragraph explaining the details).  Thanks for getting rolling on this!

laijasmine commented:
@csjx sounds good I'll work through this later today to give more detailed explanations

Commit:
remove outdated comment

Commit:
add back the slack bot check for create ticket

to accomodate changes in the RT package
now the bot checks if hte object called ticket was successfully created
if the rt_ticket_create errors, the object will not be created

Commit:
add back test now that create_ticket functionality is fixed

Commit:
added back check_rt_reply
- function now uses tryCatch to account for errors
- replaced in each spot that used to use rt_ticket_history_reply and check_rt_reply together

Commit:
removing 'create_ticket errors gracefully' test
- wasn't able to pinpoint how to get create_ticket to error
- the code is the same as the test "we can create an RT ticket" so I'm not sure how one fails and the other creates a ticket successfully

Commit:
revert exampledb.csv to original so tests will pass again

Commit:
add test for check_rt_reply

laijasmine commented:
@csjx no rush on this. I finally got back to this and brought back the `check_rt_reply` function. The difficulty is figuring how to test when it errors. I haven't figured out any good solutions to force RT to throw an error so any ideas on that would be great.

csjx commented on R/RT_functions.R:
```diff
@@ -37,17 +37,14 @@ create_ticket <- function(award, requestor) {
-    slackr::slackr_bot(out)
-    return('rt_ticket_create_error')
+  #check to see if the object ticket is created successfully
+  if(!exists("ticket")) {
```
Hi Jasmine, are you testing for the existence of the string "ticket" here? I'm not quite following that. Perhaps that should be unquoted to be the ticket variable?

laijasmine replied:
Thanks Chris! Will look at this as soon as I can.

laijasmine replied:
~~finally looked at it. Looks like it was a mistake on my part and I have fixed it in the updated commit. Thanks for catching that!~~

Me 3 hours ago wasn't thinking very clearly.

The `exists` function looks to see if the object `ticket` was created successfully. It takes a character string.

csjx replied:
gotcha! 👍

csjx reviewed (approved):
Heya Jasmine, I finally had a look at these changes. Looks good other than my one minor comment about testing for a string existence. Let's fix that and move forward on deploying this.

Commit:
ticket was mistakenly make a character in create_ticket when checking to see if the ticket was created successfully earlier using rt_ticket_creat

Commit:
remove the help in the readme for installing develop branch of RT, the fix has been merged into the main branch now

Commit:
reverting back to 0d4512d, the quotation around the ticket is needed for the exists function, removed by mistake earlier

laijasmine commented:
@csjx Need to investigate one weird problem I ran into today. Not sure why the `test_main` function is emailing the PIs when the given example database only has 2 records and should only email me. Not aware that I'm doing anything different from how I have been testing in the past or how it is outlined in the readme. Will see if I can find the cause by just tracing the code.

Commit:
update the readme so the testing procedures are more clear with step by step instructions so mistakes are less likely to be made

laijasmine commented:
Ok figured it out. Turns out I was accidentally referring to an older awards database from previous testing and was updating the awards database with new awards. I updated the readme file is clear on how to do that from the test_main.R and others in the future shouldn't run into this issue.

I'll merge this file once the checks are complete. @csjx  Let me know what the next steps should be.
